pkg/cmd/pulumi/install.go

```diff
@@ -114,60 +106,15 @@ func newInstallCmd() *cobra.Command {
 
 			if !noPlugins {
 				// Compute the set of plugins the current project needs.
-				installs, err := lang.GetRequiredPlugins(programInfo)
+				plugins, err := lang.GetRequiredPlugins(programInfo)
 				if err != nil {
 					return err
 				}
 
-				// Now for each kind, name, version pair, download it from the release website, and install it.
-				for _, install := range installs {
-					// PluginSpec.String() just returns the name and version, we want the kind too.
-					label := fmt.Sprintf("%s plugin %s", install.Kind, install)
-
-					// If the plugin already exists, don't download it unless --reinstall was passed.
-					if !reinstall {
-						if install.Version != nil {
-							if workspace.HasPlugin(install) {
-								logging.V(1).Infof("%s skipping install (existing == match)", label)
-								continue
-							}
-						} else {
-							if has, _ := workspace.HasPluginGTE(install); has {
-								logging.V(1).Infof("%s skipping install (existing >= match)", label)
-								continue
-							}
-						}
-					}
-
-					pctx.Diag.Infoerrf(diag.Message("", "%s installing"), label)
-
-					// If we got here, actually try to do the download.
-					withProgress := func(stream io.ReadCloser, size int64) io.ReadCloser {
-						return workspace.ReadCloserProgressBar(stream, size, "Downloading plugin", displayOpts.Color)
-					}
-					retry := func(err error, attempt int, limit int, delay time.Duration) {
-						pctx.Diag.Warningf(
-							diag.Message("", "Error downloading plugin: %s\nWill retry in %v [%d/%d]"), err, delay, attempt, limit)
-					}
+				pluginSet := engine.NewPluginSet(plugins...)
 
-					r, err := workspace.DownloadToFile(install, withProgress, retry)
-					if err != nil {
-						return fmt.Errorf("%s downloading from %s: %w", label, install.PluginDownloadURL, err)
-					}
-					defer func() {
-						err := os.Remove(r.Name())
-						if err != nil {
-							pctx.Diag.Warningf(
-								diag.Message("", "Error removing temporary file %s: %s"), r.Name(), err)
-						}
-					}()
-
-					payload := workspace.TarPlugin(r)
-
-					logging.V(1).Infof("%s installing tarball ...", label)
-					if err = install.InstallWithContext(ctx, payload, reinstall); err != nil {
-						return fmt.Errorf("installing %s: %w", label, err)
-					}
+				if err = engine.EnsurePluginsAreInstalled(ctx, pctx.Diag, pluginSet, pctx.Host.GetProjectPlugins(), reinstall); err != nil {
```

Comment:
One difference: Inside `engine.EnsurePluginsAreInstalled`, there is a check:

https://github.com/pulumi/pulumi/blob/487b4a8494ef556063aaf8335e6bad6317bb1072/pkg/engine/plugins.go#L219-L225

When we're doing an explicit `pulumi install`, I don't think we need to respect this envvar.